source/common/matcher/matcher.h

```diff
@@ -166,7 +166,7 @@ class MatchTreeFactory : public OnMatchFactory<DataType> {
     case MatcherType::MATCHER_TYPE_NOT_SET:
       return createAnyMatcher(config);
     }
-    return nullptr;
+    PANIC_DUE_TO_CORRUPT_ENUM;
```

Comment:
I think that panicking might not be the best approach, as a wrong config could kill the Envoy client.

I'm assuming this is only called on the config-path, and if so throwing an exception (or returning nullptr and throwing an exception) will probably be a better approach. WDYT?

Reply:
Throwing exception probably is fine. But i think `PANIC_DUE_TO_CORRUPT_ENUM` is standard approach for such proto corrupted enum situation. It is widely used in code base. For example [here](https://github.com/envoyproxy/envoy/blob/ce4d425e62cd86554c2adcbff9210dbd40c4cc3f/source/common/router/scoped_rds.cc#LL78C3-L78C28)

Adding exception code will reduce code coverage as it cannot be hit by test.

Reply:
Also, throwing the exception(and no catch)will terminate the program, which is same purpose as PANIC which will abort for critical errors like this. Basically, I think kill the Envoy is what we want for for such error.

Reply:
> Throwing exception probably is fine. But i think `PANIC_DUE_TO_CORRUPT_ENUM` is standard approach for such proto corrupted enum situation. It is widely used in code base. For example [here](https://github.com/envoyproxy/envoy/blob/ce4d425e62cd86554c2adcbff9210dbd40c4cc3f/source/common/router/scoped_rds.cc#LL78C3-L78C28)
> 
Might need to be changed there as well (depending on whether to code is executed on the data-path or config-path).

> Adding exception code will reduce code coverage as it cannot be hit by test.
Not necessarily, I think tests can still trigger these exceptions. Specifically, in the case of matcher_type, there's no requirement that validates that one of the currently 2 available types are used, so I think a test can be added that triggers the exception/panic.

> Also, throwing the exception(and no catch)will terminate the program, which is same purpose as PANIC which will abort for critical errors like this. Basically, I think killing/terminating the Envoy is what we want for for such error.
That's not true on the config-path (at least for now).
When an invalid configuration is received, the code throws an exception which is being caught, and the config is rejected.

Reply:
> I'm might be missing the validation rule, or maybe looking at the wrong field. Can you point me to the field?
[Here](https://github.com/envoyproxy/envoy/blob/4876e6911e439b68c82040a3774a37de0d3e78c0/api/envoy/config/common/matcher/v3/matcher.proto#L132) is the proto validation rule. I think the panic here is not tested as confirm by [coverage](https://storage.googleapis.com/envoy-pr/9994f66/coverage/source/common/matcher/matcher.h.gcov.html) 

> It is somewhat dependent on the subscription type that is being used. This is for [gRPC subscriptions[(https://github.com/envoyproxy/envoy/blob/a2a153d65743100367871685de7b88e4afb88256/source/extensions/config_subscription/grpc/grpc_mux_impl.cc#L291).

Thanks! I didn't know about how exception is handled in config path exactly. Since it is handled by something like NACK, it is definitely different behavior than crashinig the Envoy

> The problem we are trying to solve is 2 different binaries each seeing a different version of the proto. Let's say Envoy is "old" and has a proto that has 2 oneof options, whereas the management server is "new" and see a proto with 3 oneof options. The server may send a response with the 3rd option, and that may kill the Envoy. By throwing an exception, Envoy will send a NACK back, and continue functioning.
> 
> Not using exception is ok, as long as we are changing everything to return Status/StatusOr. There is on-going work to remove some exceptions usage (see [example](https://github.com/envoyproxy/envoy/commit/ad5f4c6cbfcc7e09fcdd09289e13ac72fcffbce4)), but using PANIC instead is not the right approach.

I think I agree that exception is better than exception/crashing the whole fleet here. But in order to throw exception  we need to add exception of throwing in header file [here](https://github.com/tyxia/envoy/blob/37f764e056505632ae12a22d215fa46305be4053/tools/code_format/config.yaml#L84)  because it is template function that can only in header filter. This is not ideal either as some products don't want exception will probably include exception throw behavior. 

That statusOr example is slightly different as it doesn't handle corrupted enum. I think it could be a future direction but I can see lots of code change as you need to handle all the caller sites. 
Besides,  I do think style guide need clear guidance or documents. PANIC indeed is widely used for corrupted enum in codebase and this "source/common/matcher/matcher.h" file alone already have 4 PANICs.

For this PR, I don't have strong preference between EITHER 1) exception approach with rule change OR 2) just panic. So it is your call .

Reply:
FWIW, panic should be used if there was a validation of the config already (this will cause a crash on the data-plane side, but should never reach that code).

I was under the impression that [this](https://github.com/cncf/xds/blob/e9ce68804cb4e64cab5a52e3c8baf840d4ff87b7/xds/type/matcher/v3/matcher.proto#L127) was the field.
I think that the file you pointed out to is the relevant one, but I'm not certain if the validation fails if 2 different proto versions of the oneof exist. That said, as long as there is a validation, it should be performed at the validation step (so maybe we'll need to fix PGV).
In light of this information, I agree with your initial approach of using panic.

Reply:
Yea that xDS proto is not the one used here, I revert back to original approach. 

Thanks for review and sharing the details about config exception handling!

PTAL